# Modern Data Structures Homework 2

Hsin-Ju (Lulu) Huang  
hh3105

In [93]:
# Import libraries
import pandas as pd
import numpy as np

In [94]:
# Import data
file_path = "C:/Users/lulu/Documents/GitHub/QMSS-GR5072-Spring2025_Class/HW/HW02/Seasons_Stats_NBA.csv"
df = pd.read_csv(file_path)

In [95]:
# Checking the imported data by showing the first and last 5 rows
display(df.head(), df.tail())

Year           Player  Pos   Age   Tm     G  GS  MP  PER    TS%  ...  \
0  1950.0  Curly Armstrong  G-F  31.0  FTW  63.0 NaN NaN  NaN  0.368  ...   
1  1950.0     Cliff Barker   SG  29.0  INO  49.0 NaN NaN  NaN  0.435  ...   
2  1950.0    Leo Barnhorst   SF  25.0  CHS  67.0 NaN NaN  NaN  0.394  ...   
3  1950.0       Ed Bartels    F  24.0  TOT  15.0 NaN NaN  NaN  0.312  ...   
4  1950.0       Ed Bartels    F  24.0  DNN  13.0 NaN NaN  NaN  0.308  ...   

     FT%  ORB  DRB  TRB    AST  STL  BLK  TOV     PF    PTS  
0  0.705  NaN  NaN  NaN  176.0  NaN  NaN  NaN  217.0  458.0  
1  0.708  NaN  NaN  NaN  109.0  NaN  NaN  NaN   99.0  279.0  
2  0.698  NaN  NaN  NaN  140.0  NaN  NaN  NaN  192.0  438.0  
3  0.559  NaN  NaN  NaN   20.0  NaN  NaN  NaN   29.0   63.0  
4  0.548  NaN  NaN  NaN   20.0  NaN  NaN  NaN   27.0   59.0  

[5 rows x 52 columns]

Year             Player Pos   Age   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller  PF  24.0  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller   C  27.0  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman   C  20.0  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser  SF  22.0  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac   C  19.0  LAL  38.0  11.0   609.0  17.0   

         TS%  ...    FT%    ORB    DRB    TRB   AST   STL   BLK   TOV     PF  \
24686  0.604  ...  0.679  135.0  270.0  405.0  99.0  62.0  58.0  65.0  189.0   
24687  0.508  ...  0.564   43.0   81.0  124.0  42.0   7.0  21.0  20.0   61.0   
24688  0.346  ...  0.600   11.0   24.0   35.0   4.0   2.0   5.0   3.0   17.0   
24689  0.503  ...  0.775   15.0  110.0  125.0  36.0  15.0  16.0  40.0   78.0   
24690  0.547  ...  0.653   41.0  118.0  159.0  30.0  14.0  33.0  30.0   66.0   

         PTS  
24686  639.0  
24687  178.0  
24688   23.0  
24689  240.0  
24690  284.0  

[5 rows x 52 columns]

In [96]:
# Get basic info about the dataset
df.info()  # Overview of data types and missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24691 entries, 0 to 24690
Data columns (total 52 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Year    24624 non-null  float64
 1   Player  24624 non-null  object 
 2   Pos     24624 non-null  object 
 3   Age     24616 non-null  float64
 4   Tm      24624 non-null  object 
 5   G       24624 non-null  float64
 6   GS      18233 non-null  float64
 7   MP      24138 non-null  float64
 8   PER     24101 non-null  float64
 9   TS%     24538 non-null  float64
 10  3PAr    18839 non-null  float64
 11  FTr     24525 non-null  float64
 12  ORB%    20792 non-null  float64
 13  DRB%    20792 non-null  float64
 14  TRB%    21571 non-null  float64
 15  AST%    22555 non-null  float64
 16  STL%    20792 non-null  float64
 17  BLK%    20792 non-null  float64
 18  TOV%    19582 non-null  float64
 19  USG%    19640 non-null  float64
 20  blanl   0 non-null      float64
 21  OWS     24585 non-null  float64
 22

In [97]:
# Check for missing values in each column
df.isnull().sum() # Summarizing missing values across columns

Year         67
Player       67
Pos          67
Age          75
Tm           67
G            67
GS         6458
MP          553
PER         590
TS%         153
3PAr       5852
FTr         166
ORB%       3899
DRB%       3899
TRB%       3120
AST%       2136
STL%       3899
BLK%       3899
TOV%       5109
USG%       5051
blanl     24691
OWS         106
DWS         106
WS          106
WS/48       590
blank2    24691
OBPM       3894
DBPM       3894
BPM        3894
VORP       3894
FG           67
FGA          67
FG%         166
3P         5764
3PA        5764
3P%        9275
2P           67
2PA          67
2P%         195
eFG%        166
FT           67
FTA          67
FT%         925
ORB        3894
DRB        3894
TRB         379
AST          67
STL        3894
BLK        3894
TOV        5046
PF           67
PTS          67
dtype: int64

## Question 1
### [5 pts] Which NBA player scored the most points in a single season? What seasons/year was this?

In [98]:
# 1. Find the NBA player who scored the most points in a single season
most_points_row = df.loc[df['PTS'].idxmax()]
most_points_row[['Player', 'PTS', 'Year']]  # Extract player name, points, and year

Player    Wilt Chamberlain*
PTS                  4029.0
Year                 1962.0
Name: 1706, dtype: object

In [99]:
# Alternative Method: to check if there are multiple players who scored the most points in a single season
max_points = df['PTS'].max()  # Get the highest points scored in a season
most_points_players = df[df['PTS'] == max_points]  # Get all players who scored that amount
most_points_players[['Player', 'PTS', 'Year']]  # Display relevant details

Player     PTS    Year
1706  Wilt Chamberlain*  4029.0  1962.0

### Wilt Chamberlain holds the record for the most points scored in a single season, with 4,029 points in 1962.

## Question 2
### [5 pts] Which NBA player scored the most points in 1991?

In [100]:
# 2. Find the NBA player who scored the most points in 1991
most_points_1991 = df[df['Year'] == 1991].loc[df[df['Year'] == 1991]['PTS'].idxmax()]
most_points_1991[['Player', 'PTS']]  # Player and points scored in 1991

Player    Michael Jordan*
PTS                2580.0
Name: 10212, dtype: object

### Michael Jordan holds the record for most points scored in 1991, with 2580 points.

## Question 3
### [10 pts] Which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data? If there are multiple players who have tied for the best percentage, then sort on the column with player names and submit your answer as a pd.DataFrame object.

In [101]:
# 3. Best free throw percentage from 2000 onward
# Filter the data for the year 2000 onward and sort by FT% (descending) and Player (ascending)
best_ft = df[df['Year'] >= 2000].sort_values(by=['FT%', 'Player'], ascending=[False, True])

# Find the highest FT% in the sorted data
max_ft = best_ft['FT%'].max()

# Filter the players with the highest FT% and return the result as a DataFrame
best_ft_players = best_ft[best_ft['FT%'] == max_ft]

#Display relevant columns of first and last 5 rows
display(best_ft_players[['Player', 'FT%', 'Year']].head(), best_ft_players[['Player', 'FT%', 'Year']].tail())


Player  FT%    Year
17990      Aaron Miles  1.0  2006.0
16181  Adam Harrington  1.0  2003.0
22787    Adonis Thomas  1.0  2014.0
22788    Adonis Thomas  1.0  2014.0
18961   Adrian Griffin  1.0  2008.0

Player  FT%    Year
17486    Wesley Person  1.0  2005.0
18271     Will Blalock  1.0  2007.0
17572      Yuta Tabuse  1.0  2005.0
17167  Zarko Cabarkapa  1.0  2005.0
16176  Zendon Hamilton  1.0  2003.0

In [102]:
best_ft_players.info()

<class 'pandas.core.frame.DataFrame'>
Index: 313 entries, 17990 to 16176
Data columns (total 52 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Year    313 non-null    float64
 1   Player  313 non-null    object 
 2   Pos     313 non-null    object 
 3   Age     313 non-null    float64
 4   Tm      313 non-null    object 
 5   G       313 non-null    float64
 6   GS      313 non-null    float64
 7   MP      313 non-null    float64
 8   PER     313 non-null    float64
 9   TS%     313 non-null    float64
 10  3PAr    313 non-null    float64
 11  FTr     313 non-null    float64
 12  ORB%    313 non-null    float64
 13  DRB%    313 non-null    float64
 14  TRB%    313 non-null    float64
 15  AST%    313 non-null    float64
 16  STL%    313 non-null    float64
 17  BLK%    313 non-null    float64
 18  TOV%    313 non-null    float64
 19  USG%    313 non-null    float64
 20  blanl   0 non-null      float64
 21  OWS     313 non-null    float64
 22  D

### From the year 2000 onwards, 313 players tied for the free throw percentage of 100%. They are listed in alphabetical order in the data frame best_ft_players.

## Question 4
### [10 pts] Among those with at least 50 attempts ('FTA'), which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data?

In [103]:
# 4. Best free throw percentage with at least 50 attempts from 2000 onward
best_ft_50 = df[(df['Year'] >= 2000) & (df['FTA'] >= 50)].sort_values(by=['FT%','Player'], ascending=[False, True])

# Find the highest FT% in the sorted data
max_ft_50 = best_ft_50['FT%'].max()

# Filter the players with the highest FT% and return the result as a DataFrame
best_ft_50_players = best_ft_50[best_ft_50['FT%'] == max_ft_50]

#Display relevant columns of first and last 5 rows
display(best_ft_50_players[['Player', 'FT%', 'FTA', 'Year']].head(1))

Player    FT%   FTA    Year
15565  Terrell Brandon  0.988  84.0  2002.0

### Among those with at least 50 attempts from the year 2000 onwards, Terrell Brandon had the best free throw percentage of 98.8%.

## Question 5
### [5 pts] Rename the variable "Pos" to "Position".

In [104]:
# 5. Rename "Pos" to "Position" (modifies the existing DataFrame directly, without the need for reassignment)
df.rename(columns={'Pos': 'Position'}, inplace=True)

In [105]:
# Alternative Method: Creates a new DataFrame with the column renamed and assigns it back to the original df variable
df = df.rename(columns={'Pos': 'Position'})

In [106]:
# Check if renamed correctly
df['Position']

0        G-F
1         SG
2         SF
3          F
4          F
        ... 
24686     PF
24687      C
24688      C
24689     SF
24690      C
Name: Position, Length: 24691, dtype: object

## Question 6
### [10 pts] Use this variable to create two variables that are called "first_position" and "second_position". Hint: If a player plays two positions, then the "-" is used as a delimiter between their first and second positions. In these cases, you will have to split the position variable into two.

If you are not familiar with the different positions in basketball, I have put a list of them below:

PF = Power Forward

SF = Strong Forward

SG = Shooting Guard

PG = Point Guard

G = Guard

C = Center

In [107]:
# 6. Split position into two new columns "first_position" and "second_position"
df[['first_position', 'second_position']] = df['Position'].str.split('-', expand=True)

In [108]:
# Check new columns
df['first_position'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 24691 entries, 0 to 24690
Series name: first_position
Non-Null Count  Dtype 
--------------  ----- 
24624 non-null  object
dtypes: object(1)
memory usage: 193.0+ KB


In [109]:
# Check new columns
df['second_position'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 24691 entries, 0 to 24690
Series name: second_position
Non-Null Count  Dtype 
--------------  ----- 
479 non-null    object
dtypes: object(1)
memory usage: 193.0+ KB


## Question 7
### [5 pts] Combine these two variables back into a single variable called "position_united". Make sure to separate the two positions by "-", as the original data field did. If there is no second position, then there will only be a single position. Again, this is similar to how the original variable was coded. Present a frequency table of "position_united".

In [110]:
# 7. Recombine these two columns into "position_united"
df['position_united'] = df[['first_position', 'second_position']].apply(lambda x: '-'.join(x.dropna()), axis=1)

# Display frequency table of "position_united"
df['position_united'].value_counts()

position_united
PF       4966
SG       4811
C        4759
SF       4699
PG       4648
G         139
F         123
F-C        81
G-F        79
           67
F-G        60
C-F        35
C-PF       33
SF-SG      30
PG-SG      28
PF-C       27
SG-PG      27
SG-SF      26
PF-SF      25
SF-PF      21
SG-PF       3
C-SF        2
SF-PG       1
PG-SF       1
Name: count, dtype: int64

## Question 8
### [10 pts] Create two new datasets.

1. a new dataset from the original dataset that includes all data except the age variable (be sure to give this dataset a new name).
2. a new dataset from the original dataset that includes the year, the player name, and age.
3. add a new column to both datasets called mergeid that includes a sequence of numbers beginning with a 1 in the first row of the data and ending with the total number of rows in the last row of the data

In [111]:
# 8. Create two new datasets (one excluding "Age", one with "Year", "Player", and "Age")
df_no_age = df.drop(columns=['Age'])  # Dataset without Age
df_age = df[['Year', 'Player', 'Age']].copy()  # Dataset with only Year, Player, and Age

In [112]:
# Add "mergeid" column to both datasets
df_no_age['mergeid'] = range(1, len(df_no_age) + 1)
df_age['mergeid'] = range(1, len(df_age) + 1)

## Question 9
### [10 pts] Join the two datasets from question (8) together to recreate the original dataset plus the new merge id.

In [113]:
# 9. Merge datasets back together
merged_df = pd.merge(df_no_age, df_age, on='mergeid')

## Question 10
### [10 pts] Subset the original dataset to the years 1995-1997. Group the data by team name and year and then summarize the average number of points per team. Arrange from most to least points. (Hint: Your answer should have a shape of 88 x 3, since we are looking at the average points scored over the 1995-1997 seasons. Note that 2 of these columns could be indexes, in which case Pandas will tell you your data is 88 rows x 1 columns).

In [120]:
# 10. Subset data for 1995-1997 and summarize average points per team
# Subset data for 1995-1997
df_95_97 = df[(df['Year'] >= 1995) & (df['Year'] <= 1997)]

# Group by team name and year, and calculate average points per team
team_average_pts = df_95_97.groupby(['Tm', 'Year'])['PTS'].mean().reset_index()

# Sort the data by average points in descending order
team_average_pts.sort_values(by='PTS', ascending=False, inplace=True)

# Show the result
team_average_pts

Tm    Year         PTS
73  SEA  1996.0  659.384615
72  SEA  1995.0  646.785714
47  MIN  1997.0  606.307692
54  ORL  1995.0  606.066667
60  PHO  1995.0  604.866667
..  ...     ...         ...
20  DEN  1997.0  348.695652
50  NJN  1997.0  346.695652
58  PHI  1996.0  322.750000
79  TOT  1997.0  322.380952
17  DAL  1997.0  275.222222

[88 rows x 3 columns]

## Question 11
### [10 pts] Similar to Q10, start by subsetting the original dataset to the years 1995-1997. Then create a wide dataset that keeps each year in its own row, but spreads team names to multiple columns with each column delineating points per team in each of the 1995-1997 seasons (Hint: Your final answer should have a shape of 3 x 30).

In [122]:
# 11. Create wide format dataset for 1995-1997 points per team
df_wide = team_average_pts.pivot(index='Year', columns='Tm', values='PTS')

In [128]:
df_wide

Tm             ATL         BOS         CHH         CHI         CLE  \
Year                                                                 
1995.0  440.055556  495.764706  515.562500  520.312500  494.466667   
1996.0  447.722222  471.944444  401.476190  575.000000  498.200000   
1997.0  457.294118  458.222222  450.444444  563.866667  551.769231   

Tm             DAL         DEN         DET         GSW         HOU  ...  \
Year                                                                ...   
1995.0  604.428571  488.764706  503.312500  481.500000  499.470588  ...   
1996.0  560.600000  471.352941  558.714286  490.235294  466.888889  ...   
1997.0  275.222222  348.695652  482.687500  510.687500  458.222222  ...   

Tm             PHO         POR         SAC         SAS         SEA  \
Year                                                                 
1995.0  604.866667  528.187500  537.066667  546.375000  646.785714   
1996.0  503.058824  543.000000  510.187500  565.133333  659.384615   
1997.0  366.565217  427.052632  494.250000  370.900000  517.125000   

Tm             TOR         TOT         UTA       VAN         WSB  
Year                                                              
1995.0         NaN  488.681818  581.733333       NaN  515.125000  
1996.0  420.736842  419.767857  600.285714  409.0000  400.380952  
1997.0  372.809524  322.380952  563.600000  457.0625  543.133333  

[3 rows x 30 columns]

## Question 12
### [10 pts] Now return the data from #11 back into a long (tidy) format by moving Team (TM), Points and Year into their own columns (Hint: Your data shape should be 90 x 3).

In [135]:
# 12. Convert wide dataset back to long format
df_long = df_wide.reset_index().melt(id_vars=['Year'], var_name='Team', value_name='PTS')

In [134]:
df_long

Year Team         PTS
0   1995.0  ATL  440.055556
1   1996.0  ATL  447.722222
2   1997.0  ATL  457.294118
3   1995.0  BOS  495.764706
4   1996.0  BOS  471.944444
..     ...  ...         ...
85  1996.0  VAN  409.000000
86  1997.0  VAN  457.062500
87  1995.0  WSB  515.125000
88  1996.0  WSB  400.380952
89  1997.0  WSB  543.133333

[90 rows x 3 columns]

In [133]:
display(df_long.head(), df_long.tail())  

Year Team         PTS
0  1995.0  ATL  440.055556
1  1996.0  ATL  447.722222
2  1997.0  ATL  457.294118
3  1995.0  BOS  495.764706
4  1996.0  BOS  471.944444

Year Team         PTS
85  1996.0  VAN  409.000000
86  1997.0  VAN  457.062500
87  1995.0  WSB  515.125000
88  1996.0  WSB  400.380952
89  1997.0  WSB  543.133333